### ISSUE TITLE
More Advanced Selection Key Bindings

### ISSUE THREAD

**Issue by mcmah309** [curlpipe/ox#172]
**What feature would you like to see? (give details)**

Ctrl+Shift+Home: Should select from the current cursor position to the start of the line
Ctrl+Shift+End: Should select from the current cursor position to the end of the line

Edit:

Per discussion below -
- [x] Shift+Home/End - select part of line

- [x] Ctrl+Shift+Left/Right - select words
- [x] Shift+PageUp/PageDown - select entire pages

**Comment by curlpipe** [comment]
This behaviour can be achieved by adding the following lua code to your configuration file (`~/.oxrc` on UNIX-based system)

```lua
event_mapping["ctrl_shift_home"] = function()
    local n_moves = editor.cursor.x
    for i = 1, n_moves do
        editor:select_left()
    end
end

event_mapping["ctrl_shift_end"] = function()
    local length = #editor:get_line()
    local n_moves = length - editor.cursor.x
    for i = 1, n_moves do
        editor:select_right()
    end
end
```

**Comment by mcmah309** [comment]
Thanks!

Hmm, `~/.oxrc` doesn't seem to work. I tried creating the file with the above sample and copy adding https://github.com/curlpipe/ox/blob/master/config/.oxrc before. Also tried specifying the file with the `-c` flag.

But yes, to me it makes sense to be a default since it also is as such on most other editors. Especially those that support Ctrl+z, Ctrl+y, Ctrl+c, Ctrl+v, pattern.

As as side: Ctrl+Shift+(page up, page down) should likely also be supported as well

**Comment by mcmah309** [comment]
Just did a bit more testing, the `ctrl_shift_*` pattern does not work. But If I changed it to something like `ctrl_*` it does work.

**Comment by curlpipe** [comment]
> Just did a bit more testing, the ctrl_shift_* pattern does not work. But If I changed it to something like ctrl_* it does work.

I see, so you're saying that `ctrl_home` / `ctrl_end` works but `ctrl_shift_home` / `ctrl_shift_end` doesn't work?

As for the ctrl+shift+pageup/pagedown, you're totally right. 

I can use this issue for a "More advanced selection key bindings" which could add:
- [ ] Ctrl+Shift+PageUp/PageDown - select entire pages
- [ ] Ctrl+Shift+Left/Right - select words
- [ ] Ctrl+Shift+Home/End - select part of line

**Comment by mcmah309** [comment]
> I see, so you're saying that `ctrl_home` / `ctrl_end` works but `ctrl_shift_home` / `ctrl_shift_end` doesn't work?

Yes



> I can use this issue for a "More advanced selection key bindings" which could add:

Perfect! I will change the issue title.

**Comment by curlpipe** [comment]
I'm trying to see if I can reproduce this issue with it not recognising ctrl_shift_home/end. Which terminal emulator are you using and on which Operating System?

**Comment by mcmah309** [comment]
I am using the `Kitty` as my terminal emulator. I just checked and there is a default binding

https://sw.kovidgoyal.net/kitty/overview/#tabs-and-windows

I ubound it with
```
map ctrl+shift+home no_op
map ctrl+shift+end no_op
```
As suggested here https://superuser.com/a/1772297 with no success

**Comment by curlpipe** [comment]
Interesting. I'll be sure to do an investigation and find out what's happening here.

**Comment by curlpipe** [comment]
I have installed kitty and added the `no_op` section to the configuration file and Ox picks up the ctrl+shift+home/end key bindings.

Could it be that you added the above lua code above the main 
```lua
event_mapping = {
...
}
```
section in the configuration file?

Try adding it at the end of the configuration file and see if that works?

**Comment by mcmah309** [comment]
I've tried both ways. Right now, only this is in my `~/.oxrc` file:
```lua
event_mapping["ctrl_shift_home"] = function()
    local n_moves = editor.cursor.x
    for i = 1, n_moves do
        editor:select_left()
    end
end

event_mapping["ctrl_d"] = function()
    local n_moves = editor.cursor.x
    for i = 1, n_moves do
        editor:select_left()
    end
end
```
The first does not work (as discussed), but `ctrl_d` does.

I have verified that my keyboard is indeed sending the right commands through other apps and https://keyboardchecker.com/ . And then I also ran `kitty --debug-keyboard` to see if kitty was receiving the keyboard input and `kitty --dump-commands` to see if kitty was executing any commands - both look good.

**Comment by mcmah309** [comment]
Wait.. looking again at the output. Looks like kitty is missing picking up the `ctrl` command. And indeed when I change the `~/.oxrc` file to `shift_home` it works. Hmm may be something I need to look into on the kitty side.

**Comment by mcmah309** [comment]
Huh well I tried another keyboard and `ctrl_shift_home` works. So this may just be a funky issue with the signals being sent from the other keyboard and kitty, since it seems to work other places with that keyboard.

But this brought up a decent point, maybe it should be just:
- Shift+PageUp/PageDown - select entire pages
- Ctrl+Shift+Left/Right - select words
- Shift+Home/End - select part of line

Dropping `Ctrl` from the 1st and the 3rd. That seems to be the current behavior on my instance of vscode as well.

**Comment by curlpipe** [comment]
Implemented these key bindings in 0.6.10 which was just released
